docs/Dialects/krnl.md

```diff
@@ -519,41 +553,84 @@ Matmul operation for a single pannel.
+operation ::= `krnl.matmul` $A `[` $aGlobalIndexMemStart `]` `,`
+              $B `[` $bGlobalIndexMemStart `]` `,`
+              $C `[` $cGlobalIndexMemStart `]` `,`
               `(` $loops `)` `,`
-              `(` $iComputeStart `,` $jComputeStart `,` $kComputeStart `)` `,`
+              `(` $iGlobalIndexComputeStart `,` $jGlobalIndexComputeStart `,`
+              $kGlobalIndexComputeStart `)` `,`
               `(` $iGlobalUB `,` $jGlobalUB `,` $kGlobalUB `)`
               attr-dict `:` type($A) `,` type($B)`,` type($C) `,` `(` type($loops) `)`
 ```
 
-Perform a matrix multiplication A * B + C
-for a small tile A * B + C of sizes
-[IxK] * [KxJ] + [IxJ].
-
-The i/j/k ComputeStarts indicate the global indices of the first element
-of a tile to be computed in the original computations.
-The i/j/k GlobalUBs indicate the upper bounds in the original computations.
+Perform a matrix multiplication A * B + C with sizes [IxK] * [KxJ] + [IxJ].
+The original matrices A, B, and C can be buffered in buffered arrays
+which may be padded. The original matrices and the padded array might
+have a higher rank than 2, but the actual matrix multiplication operation
+only deal with the innermost 2 ranks of the matrices to perform its matrix
+multiplication operations.
+
+The computations may also compute only a sub-tile of the buffered arrays.
+This region is depicted using stars '*' below.
+
+All indices passed to this operation are the global indices in the original
+computation, so as to better know if we have boundary conditions.
+
+ORIGINAL ARRAY: if A: *xIxK; if B: *xKxJ; if C: *xI*J).
+ -------------------------------------------------
+ |                                               ]
+ |                                               ]
+ |             buffer array       buffer pad     ]
+ |            (3)---------------- ++++           ]
+ |             |                 |   +           ]
+ |             |     (1)****     |   +           ]
+ |             |      *    *     |   +           ]
+ |             |      *    *     |   +           ]
+ |             |      ****(5)    |   +           ]
+ |             |                 |   +           ]
+ |             |                 |   +           ]
+ |             ------------------|   +           ]
+ |             +                     +           ]
+ |             +++++++++++++++++++++(4)          ]
+ |                                               ]
+ -----------------------------------------------(2)
+
+(1) iGlobalIndexComputeStart/jGlobalIndexComputeStart/
+    kGlobalIndexComputeStart, required, global 1D indices.
+(2) iGlobalUB/jGlobalUB/jGlobalUB, required, global 1D indices.
+(3) aGlobalIndexMemStart/bGlobalIndexMemStart/cGlobalIndexMemStart,
+    required, global nD indices with the same rank as the buffers
+    a, b, and c.
+(4) aTileSize/bTileSize/cTileSize, required when padding, 2D sizes.
+(5) computeTileSizes, required when tiled computation within buffer,
+    3D sizes (I, J, K).
```

Comment:
Really appreciated this explanation! Some times a, b, c are used instead of A, B, C, and it is a bit confusing whether A, B, C are padded buffers or original matrices.

Reply:
Thanks for the suggestion, I now use aa, bb, and cc for the buffers in the doc